sno/ogr_import_source.py

```diff
@@ -470,33 +470,64 @@ def get_geometry_v2_column_schema(self):
             ColumnSchema.new_id(), name, "geometry", None, **extra_type_info
         )
 
+    def _should_import_as_numeric(self, ogr_type, ogr_width):
+        if ogr_width == 0:
+            # Note 1: Unfortunately, these three all collide:
+            #    NUMERIC (unqualified) --> ogr.Real(0.0) --> double
+            #    FLOAT --> ogr.Real(0.0) --> double
+            #    DOUBLE PRECISION --> ogr.Real(0.0) --> double
+            # Fixing this collision might be a good reason to move away from OGR
+            # for import processing in the near future.
+            return False
+        else:
+            # This is a fixed-length numeric. OGR turns them into integers/reals,
+            # but it does actually preserve their precision, so we can turn them back.
+            return (
+                (ogr_type == "Real")
+                # when OGR is translating a modern format to a fixed-width format like shapefile,
+                # it turns int32 fields into 'Integer(9.0)' and int64 into 'Integer64(18.0)'
+                # this should generally be interpreted as an int, rather than NUMERIC(9,0)
+                or (ogr_type == "Integer" and ogr_width != 9)
```

Comment:
> where do 9 & 18 come from

these are what OGR will generate, if you ogr2ogr an INTEGER or INTEGER64 field from a sane format into a shapefile. You'll get Integer(9.0) or Integer64(18.0)

Some fiddling, starting with this postgres table:

```sql

CREATE TABLE typoes (
    bigant_pk BIGINT PRIMARY KEY,
    bigant BIGINT,
    smallant SMALLINT,
    regularant INTEGER,
    tumeric20_0 NUMERIC(20,0),
    tumeric5_5 NUMERIC(5,5),
    flote FLOAT,
    dubble DOUBLE PRECISION,
    tumeric99_0 numeric(99,0),
    tumeric4_0 numeric(4,0),
    tumeric numeric,
    techs varchar,
    techs10 varchar(100)
);
```

opening the postgres table directly with ogr shows this:
```
Layer name: typoes
Metadata:
  OLMD_FID64=YES
Geometry: None
Feature Count: 0
Layer SRS WKT:
(unknown)
FID Column = bigant_pk
bigant: Integer64 (0.0)
smallant: Integer(Int16) (5.0)
regularant: Integer (0.0)
tumeric20_0: Integer64 (20.0)
tumeric5_5: Real (5.5)
flote: Real (0.0)
dubble: Real (0.0)
tumeric99_0: Integer64 (99.0)
tumeric4_0: Integer (4.0)
tumeric: Real (0.0)
techs: String (0.0)
techs10: String (100.0)
```

notes:
- all integer[64]s with (0.0) are really ints
- the smallint has a precision but also has a subtype, so that distinguishes it from the numerics


Now, things change a bit when we ogr2ogr to DBF, and then ogrinfo the DBF:

```
bigant: Integer64 (18.0)
smallant: Integer (5.0)
regularant: Integer (9.0)
tumeric20_: Real (20.0)
tumeric5_5: Real (5.5)
flote: Real (24.15)
dubble: Real (24.15)
tumeric99_: Real (99.0)
tumeric4_0: Integer (4.0)
tumeric: Real (24.15)
techs: String (80.0)
techs10: String (100.0)
```

notes: 
- all the integer fields have a precision now (the 0.0 distinguishing actual ints has been lost)
- the original actual integers now have precisions 5 (smallint), 9 (integer), or 18 (bigint)
- other precisions are numerics

I missed the magic precision of 5 for smallint, but I can add that now. I know this stuff is awful, but this is the sanest I can make it without ripping out OGR (still should do that)

Reply:
Actually, `flote` and `double` are mishandled here too; the new code will import those as numerics because they have precision/scale. I will fix that and improve the test case.

Reply:
> This function is written slightly backwards

fair enough, i'll invert all the conditions to make it clearer

> Feels like if it's 1/2/3/4/8 bytes we should map onto a Sno integer type?

well... if it was originally an integer, we should treat it as an integer. problem is that OGR doesn't expose that so we have to guess :( the various types of sql database don't seem relevant here? obviously a better fix would be to switch to sqlalchemy rather than ogr